ignite/metrics/GAN/fid.py

```diff
@@ -0,0 +1,178 @@
+from typing import Callable, Sequence, Union
+
+import numpy as np
+import torch
+from scipy.linalg import sqrtm
+
+from ignite.exceptions import NotComputableError
+from ignite.metrics.metric import Metric, reinit__is_reduced, sync_all_reduce
+
+__all__ = ["FID", "InceptionExtractor"]
+
+
+def fid_score(
+    mu1: torch.Tensor, mu2: torch.Tensor, sigma1: torch.Tensor, sigma2: torch.Tensor, eps: float = 1e-6
+) -> float:
+
+    diff = mu1 - mu2
+
+    # Product might be almost singular
+    covmean, _ = sqrtm(sigma1.mm(sigma2), disp=False)
+    if not np.isfinite(covmean).all():
+        offset = np.eye(sigma1.shape[0]) * eps
+        covmean = sqrtm((sigma1 + offset).dot(sigma2 + offset))
+
+    # Numerical error might give slight imaginary component
+    if np.iscomplexobj(covmean):
+        if not np.allclose(np.diagonal(covmean).imag, 0, atol=1e-3):
+            m = np.max(np.abs(covmean.imag))
+            raise ValueError("Imaginary component {}".format(m))
+        covmean = covmean.real
+
+    tr_covmean = np.trace(covmean)
+
+    return diff.dot(diff) + np.trace(sigma1) + np.trace(sigma2) - 2 * tr_covmean
+
+
+class InceptionExtractor:
+    def __init__(self) -> None:
+        from torchvision import models
+
+        self.model = models.inception_v3(init_weights=True)
+        self.model.fc = torch.nn.Sequential()
+        self.model.eval()
+
+    def __call__(self, data: torch.Tensor) -> torch.Tensor:
+        if data.shape[1:] != (3, 299, 299):
+            raise ValueError(f"Images should be of size 3x299x299 (got {data.shape})")
+        return self.model(data).detach()
+
+
+class FID(Metric):
+    r"""Calculates Frechet Inception Distance.
+
+    .. math::
+       \text{FID} = \mod{\text{mu1} - \text{mu2}} + \text{Trace(sigma1 + sigma2 - 2*sqrt(sigma1*sigma2))}
+
+    where :math:`mu1` and :math:`sigma1` refer to the mean and covariance of the train data and
+    :math:`mu2` and :math:`sigma2` refer to the mean and covariance of the test data.
+
+    More details can be found in `Heusel et al. 2002`__.
+
+    __ https://arxiv.org/pdf/1706.08500.pdf
+
+    In addition, a faster and online computation approach can be found in `Chen et al. 2014`__
+
+    __ https://arxiv.org/pdf/2009.14075.pdf
+
+    Remark:
+
+        This implementation is inspired by pytorch_fid package.
+
+    Args:
+        num_features: specifies number of features the evaluation samples should have
+        eps: small value added during matric division if denominator is 0.
```

Comment:
Do we really need to specify that as an argument ? Can't we just keep it internally ?

Reply:
It was previously done lazily but @sdesrozis and I think that we should have it more standardized.

Reply:
My point is about why user should care about that ? Do we have a use-case where we need to set to `eps` another value ?
A standardized approach will be to have in somewhere in `ignite.eps` and reuse everywhere but not as class argument.

Reply:
My bad, I thought you were talking about num_features, we can have a global eps and just call ignite.eps. Point of eps is it should be negligible compared to the values of our data.

Reply:
I agree that `eps` is not relevant for the users. We should remove this parameter and use according to a default fixed value inside.